lambdas/report-user-identity/src/main/java/uk/gov/di/ipv/core/reportuseridentity/ReportUserIdentityHandler.java

```diff
@@ -0,0 +1,335 @@
+        this.reportUserIdentityScanDynamoDataStore = reportUserIdentityScanDynamoDataStore;
+        this.reportSummaryScanDynamoDataStore = reportSummaryScanDynamoDataStore;
+    }
+
+    @SuppressWarnings({"unused", "java:S107"}) // Used by AWS
+    public ReportUserIdentityHandler() {
+        this.objectMapper = new ObjectMapper();
+        this.configService = ConfigService.create();
+        this.userIdentityService = new UserIdentityService(configService);
+        this.verifiableCredentialService = new VerifiableCredentialService(configService);
+        this.reportUserIdentityService = new ReportUserIdentityService();
+        this.vcStoreItemScanDynamoDataStore =
+                new ScanDynamoDataStore<>(
+                        configService.getEnvironmentVariable(
+                                EnvironmentVariable.USER_ISSUED_CREDENTIALS_TABLE_NAME),
+                        VcStoreItem.class,
+                        DynamoDataStore.getClient(),
+                        configService);
+        this.reportUserIdentityScanDynamoDataStore =
+                new ScanDynamoDataStore<>(
+                        configService.getEnvironmentVariable(
+                                EnvironmentVariable.REPORT_USER_IDENTITY_TABLE_NAME),
+                        ReportUserIdentityItem.class,
+                        DynamoDataStore.getClient(),
+                        configService);
+        this.reportSummaryScanDynamoDataStore =
+                new ScanDynamoDataStore<>(
+                        configService.getEnvironmentVariable(
+                                EnvironmentVariable.REPORT_SUMMARY_TABLE_NAME),
+                        ReportSummaryItem.class,
+                        DynamoDataStore.getClient(),
+                        configService);
+    }
+
+    @Override
+    @Tracing
+    @Logging(clearState = true)
+    public void handleRequest(InputStream inputStream, OutputStream outputStream, Context context)
+            throws IOException {
+        LogHelper.attachComponentId(configService);
+        LOGGER.info(LogHelper.buildLogMessage("Start processing report."));
+
+        ReportProcessingResult.ReportProcessingResultBuilder reportProcessingResult =
+                ReportProcessingResult.builder();
+        try {
+            ReportProcessingRequest reportProcessingRequest =
+                    objectMapper.readValue(inputStream, ReportProcessingRequest.class);
+            LOGGER.info(
+                    LogHelper.buildLogMessage(
+                            "Request received- " + reportProcessingRequest.toString()));
+            // Step-1
+            scanToExtractUniqueUserIdFromTacticalStore(
+                    reportProcessingResult, reportProcessingRequest, context);
+            // Step-2
+            processUsersToFindLOCAndUpdateDb(
+                    reportProcessingRequest, reportProcessingResult, context);
+            // Step-3
```

Comment:
I was thinking the three steps would be:

1. Scan VC table to collect user ids (as you've done)
    - Medium speed - needs to scan all 2.5M VCs and update the other table
2. Scan user table and update with identity details
    - Slow speed - needs to fetch all VCs for each user in turn, calculate identity and store the result
3. Scan user table and aggregate identity details
    - Fast(ish) speed - needs to scan 0.5-1M user records and aggregate the data

Perhaps it's worth revising the input class to reflect these three steps, so you can choose which to execute?

At the moment I think we've also pushed some of 3 into 2 - with the extra report table. I'm not sure if this is useful, it slightly depends how quick step 3 is with another full table scan.

Reply:
Step-2 we are aggregating result and pesisting it to result table which is used in step-3 to avoid full table scan of all unique users frm report-user-identity table.